src/components/darkmode-switch/darkmode-switch.tsx

```diff
@@ -0,0 +1,132 @@
+        } as unknown as MediaQueryList;
+
+        this.colorSchemeMediaQuery =
+            (window.matchMedia &&
+                window.matchMedia('(prefers-color-scheme: dark)')) ||
+            colorSchemeMediaQueryFallback;
+    }
+
+    public connectedCallback() {
+        this.colorSchemeMediaQuery.addEventListener(
+            'change',
+            this.handleSystemThemeChange
+        );
+    }
+
+    public disconnectedCallback() {
+        this.colorSchemeMediaQuery.removeEventListener(
+            'change',
+            this.handleSystemThemeChange
+        );
+    }
+
+    public componentWillLoad(): void {
+        this.systemSettingIsDark = this.colorSchemeMediaQuery.matches;
+
+        this.setTheme(
+            (localStorage.getItem(LOCALSTORAGE_KEY) || DEFAULT) as Theme
+        );
+    }
+
+    public render(): HTMLElement {
+        const props = {
+            checked:
+                this.theme === DARK ||
+                (this.theme === DEFAULT && this.systemSettingIsDark),
+        };
+
+        return (
+            <div class="mode-toggle">
+                <input
+                    type="checkbox"
+                    onChange={this.handleThemeChange}
+                    ref={this.getSelectRef}
+                    {...props}
+                />
+                <div class="mode-visualization">
+                    <div class="circle"></div>
+                    <div class="ray one"></div>
+                    <div class="ray two"></div>
+                    <div class="ray three"></div>
+                    <div class="ray four"></div>
+                </div>
+            </div>
+        );
+    }
+
+    private getSelectRef = (element: HTMLInputElement) => {
+        this.checkbox = element;
+    };
+
+    private handleSystemThemeChange = (e: MediaQueryListEvent) => {
+        this.systemSettingIsDark = !!e.matches;
+        if (this.theme === DEFAULT) {
+            this.checkbox.checked = !this.checkbox.checked;
+        }
+    };
+
+    private handleThemeChange = () => {
+        const checkboxValue = !!this.checkbox.checked;
+
+        let newTheme: Theme;
+
+        if (this.systemSettingIsDark) {
+            if (checkboxValue === CHECKBOX_LIGHT) {
+                newTheme = LIGHT;
+            } else {
+                newTheme = DEFAULT;
+            }
+        } else {
+            if (checkboxValue === CHECKBOX_DARK) {
+                newTheme = DARK;
+            } else {
+                newTheme = DEFAULT;
+            }
+        }
+
+        this.setTheme(newTheme);
+        document.dispatchEvent(
+            new CustomEvent<Theme>(THEME_EVENT_NAME, { detail: newTheme })
+        );
+    };
+
+    private setTheme = (value: Theme) => {
+        this.theme = value;
+        document.querySelector('html').dataset.theme = value;
+        localStorage.setItem(LOCALSTORAGE_KEY, value);
+    };
+}
+
+export type Theme = 'system-default' | 'force-light' | 'force-dark';
```

Comment:
I thought Stencil complained if there is more than one export from a file :thinking: Has that been fixed? :slightly_smiling_face:

Reply:
I think it complains if there are more than one component in the same file, and maybe it used to complain if there were other exports in the same file as a component? 🤔

It doesn't complain about this anyway…

Reply:
Ah, no. I tried moving the `export const THEME_EVENT_NAME = 'kompendiumTheme';` in here, and then the build failed with the following message:

```
[ ERROR ]  ./src/components/darkmode-switch/darkmode-switch.tsx:4:14
           To allow efficient bundling, modules using
           @Component() can only have a single export which
           is the component class itself. Any other exports
           should be moved to a separate file. For further
           information check out:
           https://stenciljs.com/docs/module-bundling

      L3:  export type Theme = 'system-default' | 'force-light' | 'force-dark';
      L4:  export const THEME_EVENT_NAME = 'kompendiumTheme'

[48:22.0]  build failed in 3.82
```

So it probably just worked because this was only a type, and not a value, so it'll vanish when compiled to js. I'll move this to the other file though.

Reply:
⚡